### DISCUSSION TITLE
Create new PyObject from std::type_info and pointer

### DISCUSSION THREAD

**Discussion by noahbkim** [wjakob/nanobind#140] [Q&A] [answered]
Hi all, I'm in an odd situation where I have the `std::type_info` of a primitive/STL object (e.g. `int`, `float`, `std::string`) and an untyped pointer to it (`char*`). I would like to copy/cast this data into a PyObject to be returned by a bound function. 

I imagine this is possible with existing machinery, but I'm having trouble figuring out how! I'd greatly appreciate any help :D

I've gotten as far as the following:

```cpp
nanobind::object cast_value(
    const std::type_info* type_info,
    void* address,
    nanobind::rv_policy policy = nanobind::rv_policy::copy
)
{
    nanobind::handle handle = nanobind::detail::nb_type_put(type_info, address, policy, nullptr);
    if (!handle.is_valid())
    {
        nanobind::detail::raise_cast_error();
    }
    return nanobind::steal(handle);
}
```

When provided with the `std::type_info` from `unsigned int` and a valid pointer to one, `handle.is_valid()` is false--maybe I'm missing something about how `nb_type_put` should be used.

Followup: it seems like at the time of casting, `std::type_index(typeid(unsigned int))` is not in `internals.type_c2p`. I'll try instantiating a static caster for each type I'd expect to show up.

Followup: oh, I guess `internals.type_c2p` is for registered classes--primitives are handled by a custom caster (which doesn't seem to be accessible via `std::type_info`. So I guess the best route is to switch on the type info...

**Comment by wjakob** [comment, marked as answer]
If it's primitive types, your best bet will be to iterate through all of the possibilities, `static_cast` in C++, and then call `nb::cast`. The nanobind data structures would only be helpful for types exposed using bindings (i.e. via `nb::class_<T>`), which isn't even possible for `int` etc.

**Reply by noahbkim** [reply]
Thanks for the quick response!
